src/vm/core_test.zig

```diff
@@ -2912,3 +2914,148 @@ test "CairoVM: getReturnValues should return a memory error when Ap is 0" {
+
+    try vm.segments.memory.setUpMemory(
+        allocator,
+        .{
+            .{ .{ 2, 0 }, .{12} },
+            .{ .{ 2, 1 }, .{10} },
+            .{ .{ 2, 2 }, .{8} },
+        },
+    );
+    defer vm.segments.memory.deinitData(allocator);
+
+    try expectEqual(void, @TypeOf(try vm.verifyAutoDeductionsForAddr(allocator, Relocatable.init(2, 0), &builtin)));
+    try expectEqual(void, @TypeOf(try vm.verifyAutoDeductionsForAddr(allocator, Relocatable.init(2, 1), &builtin)));
+    try expectEqual(void, @TypeOf(try vm.verifyAutoDeductionsForAddr(allocator, Relocatable.init(2, 2), &builtin)));
+}
+
+test "CairoVM: verifyAutoDeductionsForAddr throws InconsistentAutoDeduction" {
+    const allocator = std.testing.allocator;
+
+    var bitwise_builtin = BitwiseBuiltinRunner.initDefault();
+    bitwise_builtin.base = 2;
+    const builtin = BuiltinRunner{ .Bitwise = bitwise_builtin };
+
+    var vm = try CairoVM.init(allocator, .{});
+    defer vm.deinit();
+
+    try vm.segments.memory.setUpMemory(
+        allocator,
+        .{
+            .{ .{ 2, 0 }, .{12} },
+            .{ .{ 2, 1 }, .{10} },
+            .{ .{ 2, 2 }, .{7} },
+        },
+    );
+    defer vm.segments.memory.deinitData(allocator);
+
+    try expectError(CairoVMError.InconsistentAutoDeduction, vm.verifyAutoDeductionsForAddr(allocator, Relocatable.init(2, 2), &builtin));
+}
+
+test "CairoVM: verifyAutoDeductions for bitwise builtin runner" {
+    const allocator = std.testing.allocator;
+
+    var bitwise_builtin = BitwiseBuiltinRunner.initDefault();
+    bitwise_builtin.base = 2;
+    const builtin = BuiltinRunner{ .Bitwise = bitwise_builtin };
+
+    var vm = try CairoVM.init(allocator, .{});
+    defer vm.deinit();
+
+    try vm.builtin_runners.append(builtin);
+
+    try vm.segments.memory.setUpMemory(
+        allocator,
+        .{
+            .{ .{ 2, 0 }, .{12} },
+            .{ .{ 2, 1 }, .{10} },
+            .{ .{ 2, 2 }, .{8} },
+        },
+    );
+    defer vm.segments.memory.deinitData(allocator);
+
+    const result = try vm.verifyAutoDeductions(allocator);
+    try expectEqual(void, @TypeOf(result));
+}
+
+test "CairoVM: verifyAutoDeductions for bitwise builtin runner throws InconsistentAutoDeduction" {
+    const allocator = std.testing.allocator;
+
+    var bitwise_builtin = BitwiseBuiltinRunner.initDefault();
+    bitwise_builtin.base = 2;
+    const builtin = BuiltinRunner{ .Bitwise = bitwise_builtin };
+
+    var vm = try CairoVM.init(allocator, .{});
+    defer vm.deinit();
+
+    try vm.builtin_runners.append(builtin);
+
+    try vm.segments.memory.setUpMemory(
+        allocator,
+        .{
+            .{ .{ 2, 0 }, .{12} },
+            .{ .{ 2, 1 }, .{10} },
+            .{ .{ 2, 2 }, .{7} },
+        },
+    );
+
+    defer vm.segments.memory.deinitData(allocator);
+
+    try expectError(CairoVMError.InconsistentAutoDeduction, vm.verifyAutoDeductions(allocator));
+}
+
+test "CairoVM: verifyAutoDeductions for keccak builtin runner" {
```

Comment:
Here something like:

```zig
test "CairoVM: verifyAutoDeductions for keccak builtin runner" {
    const allocator = std.testing.allocator;

    var keccak_instance_def = try KeccakInstanceDef.default(allocator);

    var vm = try CairoVM.init(allocator, .{});
    defer vm.deinit();

    try vm.segments.memory.setUpMemory(
        allocator,
        .{
            .{ .{ 0, 16 }, .{43} },
            .{ .{ 0, 17 }, .{199} },
            .{ .{ 0, 18 }, .{0} },
            .{ .{ 0, 19 }, .{0} },
            .{ .{ 0, 20 }, .{0} },
            .{ .{ 0, 21 }, .{0} },
            .{ .{ 0, 22 }, .{0} },
            .{ .{ 0, 23 }, .{1} },
            .{ .{ 0, 24 }, .{564514457304291355949254928395241013971879337011439882107889} },
            .{ .{ 0, 25 }, .{1006979841721999878391288827876533441431370448293338267890891} },
            .{ .{ 0, 26 }, .{811666116505725183408319428185457775191826596777361721216040} },
        },
    );
    defer vm.segments.memory.deinitData(allocator);

    try vm.builtin_runners.append(.{ .Keccak = KeccakBuiltinRunner.init(
        allocator,
        &keccak_instance_def,
        true,
    ) });

    const result = try vm.verifyAutoDeductions(allocator);

    try expectEqual(void, @TypeOf(result));
}
```

should be good. The main problem you had here is that even if you define your variable `keccak_builtin`, then you add it inside `vm.builtin_runners` and therefore you have to deinit inside.

In https://github.com/keep-starknet-strange/ziggy-starkdust/pull/295 I added an automatic way to deinit builtins individually in `defer vm.deinit();` call.